pylsp/plugins/definition.py

```diff
@@ -1,22 +1,49 @@
 # Copyright 2017-2020 Palantir Technologies, Inc.
 # Copyright 2021- Python Language Server Contributors.
-
+from __future__ import annotations
 import logging
+from typing import Any, Dict, List, TYPE_CHECKING
 from pylsp import hookimpl, uris, _utils
 
+if TYPE_CHECKING:
+    from jedi.api import Script
+    from jedi.api.classes import Name
+    from pylsp.config.config import Config
+    from pylsp.workspace import Document
+
 log = logging.getLogger(__name__)
 
 
+def _resolve_definition(
+    maybe_defn: Name, script: Script, settings: Dict[str, Any]
+) -> Name:
+    while not maybe_defn.is_definition() and maybe_defn.module_path == script.path:
```

Comment:
Is there a guarantee that this will never become an infinite loop, for example if `follow_imports` is True and the two modules (wrongly) import the same symbol from each other? If not, is it worth adding a limit of how many iterations can be performed, setting it to something arbitrarily large and raising an informative exception if it fails?

Reply:
Right now this doesn't follow hops across modules (that should be fixed as a followup potentially), but generally I agree it would be good to be robust to jedi or other weird user-introduced circular references. Added what feels like a reasonable limit to me (100 hops) but happy to adjust if there's another good reason to do so.